apps/emqx_web_hook/src/emqx_web_hook_actions.erl

```diff
@@ -335,8 +337,26 @@ method(POST) when POST == <<"POST">>; POST == <<"post">> -> post;
 method(PUT) when PUT == <<"PUT">>; PUT == <<"put">> -> put;
 method(DEL) when DEL == <<"DELETE">>; DEL == <<"delete">> -> delete.
 
-headers(Headers) ->
-    emqx_http_lib:normalise_headers(maps:to_list(Headers)).
+preproc_and_normalise_headers(Headers) ->
+    maps:fold(fun(K0, V, Acc) ->
+        case emqx_rule_utils:if_contains_placeholder(K0) of
+            false ->
+                %% see emqx_http_lib:normalise_headers/1 for more info
+                K = re:replace(K0, "_", "-", [{return, binary}]),
+                LK = string:lowercase(K),
+                Acc#{emqx_rule_utils:preproc_tmpl(LK) => emqx_rule_utils:preproc_tmpl(V)};
+            true ->
+                %% We don't normalize the header key if it contains any placeholder,
+                %% instead we rely on the user to ensure the correctness of key format
```

Comment:
header normalization should be done after the template rendering.
```suggestion
```

Reply:
Then we have to do the header normalization at runtime.